Mathlib/NumberTheory/ADEInequality.lean

```diff
@@ -0,0 +1,286 @@
+  admissible_E'5
+#align ADE_inequality.admissible_E8 ADEInequality.admissible_E8
+
+theorem Admissible.one_lt_sumInv {pqr : Multiset ℕ+} : Admissible pqr → 1 < sumInv pqr := by
+  rw [Admissible]
+  rintro (⟨p', q', H⟩ | ⟨n, H⟩ | H | H | H)
+  · rw [← H, A', sumInv_pqr, add_assoc]
+    simp only [lt_add_iff_pos_right, PNat.one_coe, inv_one, Nat.cast_one]
+    apply add_pos <;> simp only [PNat.pos, Nat.cast_pos, inv_pos]
+  · rw [← H, D', sumInv_pqr]
+    conv_rhs => simp only [OfNat.ofNat, PNat.mk_coe]
+    norm_num
+  all_goals
+    rw [← H, E', sumInv_pqr]
+    conv_rhs => simp only [OfNat.ofNat, PNat.mk_coe]
+
+#align ADE_inequality.admissible.one_lt_sum_inv ADEInequality.Admissible.one_lt_sumInv
+
+theorem lt_three {p q r : ℕ+} (hpq : p ≤ q) (hqr : q ≤ r) (H : 1 < sumInv {p, q, r}) : p < 3 := by
+  have h3 : (0 : ℚ) < 3 := by norm_num
+  contrapose! H
+  rw [sumInv_pqr]
+  have h3q := (not_lt.mp H).trans hpq
+  have h3r := h3q.trans hqr
+  simp at H
+  have hp: (p : ℚ)⁻¹ ≤ 3⁻¹ := by
+    rw [inv_le_inv _ h3]
+    assumption_mod_cast
+    norm_num
+  have hq: (q : ℚ)⁻¹ ≤ 3⁻¹ := by
+    rw [inv_le_inv _ h3]
+    assumption_mod_cast
+    norm_num
+  have hr: (r : ℚ)⁻¹ ≤ 3⁻¹ := by
+    rw [inv_le_inv _ h3]
+    assumption_mod_cast
+    norm_num
+  calc
+    (p : ℚ)⁻¹ + (q : ℚ)⁻¹ + (r : ℚ)⁻¹ ≤ 3⁻¹ + 3⁻¹ + 3⁻¹ := add_le_add (add_le_add hp hq) hr
+    _ = 1 := by norm_num
+#align ADE_inequality.lt_three ADEInequality.lt_three
+
+theorem lt_four {q r : ℕ+} (hqr : q ≤ r) (H : 1 < sumInv {2, q, r}) : q < 4 := by
+  have h4 : (0 : ℚ) < 4 := by norm_num
+  contrapose! H
+  rw [sumInv_pqr]
+  have h4r := (not_lt.mp H).trans hqr
+  simp at H
+  have hq: (q : ℚ)⁻¹ ≤ 4⁻¹ := by
+    rw [inv_le_inv _ h4]
+    assumption_mod_cast
+    norm_num
+  have hr: (r : ℚ)⁻¹ ≤ 4⁻¹ := by
+    rw [inv_le_inv _ h4]
+    assumption_mod_cast
+    norm_num
+  calc
+    (2⁻¹ + (q : ℚ)⁻¹ + (r : ℚ)⁻¹) ≤ 2⁻¹ + 4⁻¹ + 4⁻¹ := add_le_add (add_le_add le_rfl hq) hr
+    _ = 1 := by norm_num
+#align ADE_inequality.lt_four ADEInequality.lt_four
+
+theorem lt_six {r : ℕ+} (H : 1 < sumInv {2, 3, r}) : r < 6 := by
+  have h6 : (0 : ℚ) < 6 := by norm_num
+  contrapose! H
+  rw [sumInv_pqr]
+  simp at H
+  have hr: (r : ℚ)⁻¹ ≤ 6⁻¹ := by
+    rw [inv_le_inv _ h6]
+    assumption_mod_cast
+    norm_num
+  calc
+    (2⁻¹ + 3⁻¹ + (r : ℚ)⁻¹ : ℚ) ≤ 2⁻¹ + 3⁻¹ + 6⁻¹ := add_le_add (add_le_add le_rfl le_rfl) hr
+    _ = 1 := by norm_num
+#align ADE_inequality.lt_six ADEInequality.lt_six
+
+theorem admissible_of_one_lt_sumInv_aux' {p q r : ℕ+} (hpq : p ≤ q) (hqr : q ≤ r)
+    (H : 1 < sumInv {p, q, r}) : Admissible {p, q, r} := by
+  have hp3 : p < 3 := lt_three hpq hqr H
+  -- porting note: currently cannot work, see line 287 in IntervalCases.lean
+  interval_cases p
+  · exact admissible_A' q r
+  have hq4 : q < 4 := lt_four hqr H
+  interval_cases
+  · exact admissible_D' r
+  have hr6 : r < 6 := lt_six H
+  interval_cases
+  · exact admissible_E6
+  · exact admissible_E7
+  · exact admissible_E8
```

Comment:
If `norm_num` should fail to support `ℕ+`, try this.
```suggestion
theorem admissible_of_one_lt_sumInv_aux' {p q r : ℕ+} (hpq : p ≤ q) (hqr : q ≤ r)
    (H : 1 < sumInv {p, q, r}) : Admissible {p, q, r} := by
  have hp3 : p < 3 := lt_three hpq hqr H
  -- Porting note: `interval_cases` doesn't support `ℕ+` yet.
  replace hp3 := Finset.mem_Iio.mpr hp3
  conv at hp3 => change p ∈ ({1, 2} : Multiset ℕ+)
  fin_cases hp3
  · exact admissible_A' q r
  have hq4 : q < 4 := lt_four hqr H
  replace hq4 := Finset.mem_Ico.mpr ⟨hpq, hq4⟩; clear hpq
  conv at hq4 => change q ∈ ({2, 3} : Multiset ℕ+)
  fin_cases hq4
  · exact admissible_D' r
  have hr6 : r < 6 := lt_six H
  replace hr6 := Finset.mem_Ico.mpr ⟨hqr, hr6⟩; clear hqr
  conv at hr6 => change r ∈ ({3, 4, 5} : Multiset ℕ+)
  fin_cases hr6
  · exact admissible_E6
  · exact admissible_E7
  · exact admissible_E8
```

Reply:
It seems that handling `ℕ+` by `norm_num` is difficult, so I commited this.

Reply:
great! No need for interval_cases anymore

Reply:
Is this ready for review then, @Komyyy?